File: apps/backend/src/main/kotlin/com/lyra/app/forms/application/update/UpdateFormCommandHandler.kt
package com.lyra.app.forms.application.update

import com.lyra.app.forms.domain.FormId
import com.lyra.app.forms.domain.dto.FormStyleConfiguration
import com.lyra.app.organization.domain.OrganizationId
import com.lyra.common.domain.Service
import com.lyra.common.domain.bus.command.CommandHandler
import org.slf4j.LoggerFactory

/**
 * Service class responsible for handling form update commands.
 *
 * @property formCreator The service for updating forms.
 */
@Service
class UpdateFormCommandHandler(
    private val formCreator: FormUpdater
) : CommandHandler<UpdateFormCommand> {

    /**
     * Handles a form update command.
     * Logs the name of the form being updated, creates a FormDTO from the command, and calls the form updater service.
     *
     * @param command The form update command to handle.
     */
    override suspend fun handle(command: UpdateFormCommand) {
        log.debug("Updating form with name: ${command.name}")
        val formStyleConfiguration = FormStyleConfiguration(
            name = command.name,
            header = command.header,
            description = command.description,
            inputPlaceholder = command.inputPlaceholder,
            buttonText = command.buttonText,
            buttonColor = command.buttonColor,
            backgroundColor = command.backgroundColor,
            textColor = command.textColor,
            buttonTextColor = command.buttonTextColor,
        )
        formCreator.update(
            OrganizationId(command.organizationId),
            FormId(command.id), formStyleConfiguration,
        )
    }

    companion object {
        private val log = LoggerFactory.getLogger(UpdateFormCommandHandler::class.java)
    }
}

File: apps/frontend/packages/api-services/src/forms/application/update/UpdateFormCommandHandler.ts
import type { CommandHandler } from '@lyra/shared';
import { inject, injectable } from 'inversify';
import UpdateFormCommand from '~/forms/application/update/UpdateFormCommand.ts';
import FormUpdater from '~/forms/application/update/FormUpdater.ts';
import type { Form } from '~/forms/domain/Form.ts';
import { FORM_UPDATER_PROVIDER } from '~/di/forms/forms.module.types.ts';

/**
 * @class UpdateFormCommandHandler
 *
 * @description A class that implements the CommandHandler interface for the UpdateFormCommand.
 * It uses dependency injection to get an instance of the FormUpdater.
 */
@injectable()
export default class UpdateFormCommandHandler implements CommandHandler<UpdateFormCommand> {
	/**
	 * @constructor
	 *
	 * @param {FormUpdater} formUpdater - An instance of FormUpdater.
	 *
	 * @description The constructor uses inversify's @inject decorator to inject
	 * an instance of FormUpdater.
	 */
	constructor(@inject(FORM_UPDATER_PROVIDER) private readonly formUpdater: FormUpdater) {}

	/**
	 * @method handle
	 *
	 * @param {UpdateFormCommand} command - The command to handle.
	 *
	 * @returns {Promise<void>}
	 *
	 * @description This method handles the UpdateFormCommand by creating a form object
	 * from the command and then calling the update method on the formUpdater instance.
	 */
	async handle(command: UpdateFormCommand): Promise<void> {
		const form = {
			id: command.id,
			name: command.name,
			header: command.header,
			description: command.description,
			inputPlaceholder: command.inputPlaceholder,
			buttonText: command.buttonText,
			buttonColor: command.buttonColor,
			backgroundColor: command.backgroundColor,
			textColor: command.textColor,
			buttonTextColor: command.buttonTextColor,
		} as Form;
		await this.formUpdater.update(form);
	}
}
